### PR TITLE
Allow specificaiton of a file that contains several src/dst pairs.

### PR DESCRIPTION
**Pull request by skcoulter**
These can be lids or guids, in non-sequential order.
The use case is allowing the creation of large route maps in a significantly shorter period of time.
Testing a route map for n=128 ports, (n*(n-1)) = 16256 distinct routes , resulted in almost 7x speed up.

Signed-off-by: Susan K. Coulter <markus@wf-fe1.lanl.gov>

### CODE REVIEW

File: src/ibtracert.c [hunk 1]
```diff
@@ -811,9 +826,36 @@ int main(int argc, char **argv)
 	argc -= optind;
 	argv += optind;
 
-	if (argc < 2)
+	if (argc < 2 && ports_file == NULL) 
 		ibdiag_show_usage();
 
+	if (argc == 2) {
+		srcid = strdup(argv[0]);
+		if (srcid == 0)
```
Comment (hnrose) [inline src/ibtracert.c]: NULL r.t. 0 ?

File: src/ibtracert.c [hunk 2]
```diff
@@ -811,9 +826,36 @@ int main(int argc, char **argv)
...
+			IBEXIT("out of memory, strdup for srcid failed");
+		dstid = strdup(argv[1]);
+		if (dstid == 0)
```
Comment (hnrose) [inline src/ibtracert.c]: NULL r.t. 0 ?

File: src/ibtracert.c [hunk 3]
```diff
@@ -825,57 +867,83 @@ int main(int argc, char **argv)
 
 	node_name_map = open_node_name_map(node_name_map_file);
 
-	if (resolve_portid_str(ibd_ca, ibd_ca_port, &src_portid, argv[0],
+	while (srcid[0] != '\0' && dstid != '\0') {
```
Comment (hnrose) [inline src/ibtracert.c]: dstid[0] r.t. dstid ?

File: src/ibtracert.c [hunk 4]
```diff
@@ -825,57 +867,83 @@ int main(int argc, char **argv)
...
-	/* dump multicast path */
-	dump_mcpath(endnode, dumplevel);
+			/* dump multicast path */
+			dump_mcpath(endnode, dumplevel);
+		}
 
+		if (ports_fd) {
+			my_portid.drpath.cnt = 0;
+			src_portid.drpath.cnt = 0;
+     	   		if (fgets(ports_buffer, sizeof(ports_buffer), ports_fd)) {
```
Comment (hnrose) [inline src/ibtracert.c]: Extra tab before if ?

File: src/ibtracert.c [hunk 5]
```diff
@@ -825,57 +867,83 @@ int main(int argc, char **argv)
...
+				num_port_pairs++;
+                		if ( sscanf(ports_buffer, "%s %s", srcid, dstid) != 2 )
```
Comment (hnrose) [inline src/ibtracert.c]: Extra tab before if ?

File: src/ibtracert.c [hunk 6]
```diff
@@ -825,57 +867,83 @@ int main(int argc, char **argv)
...
+	}
+	
+/*
+ * end of loop
+*/
 	close_node_name_map(node_name_map);
 
 	mad_rpc_close_port(srcport);
 
+
```
Comment (hnrose) [inline src/ibtracert.c]: No need to add this blank line

File: src/ibtracert.c [hunk 7]
```diff
@@ -825,57 +867,83 @@ int main(int argc, char **argv)
...
 	exit(0);
 }
+
```
Comment (hnrose) [inline src/ibtracert.c]: No need to add this blank line.

**Review by hnrose — CHANGES_REQUESTED** [review]
Just some nits...

**Comment by hnrose** [comment]
Commit is titled "Allow specificaiton of a file that contains several src/dst pairs"
Better as
ibtracert: Allow specification of a file that contains several src/dst pairs

You can update the patch description at the same time too ;-)

File: src/ibtracert.c (hunk 1 above)
Comment (weiny2) [reply src/ibtracert.c]: Yes please use NULL

File: doc/rst/common/sec_ports-file.rst [hunk 8]
```diff
@@ -0,0 +1,27 @@
+.. Common text to describe the port file.
+
+PORTS FILE FORMAT
+-------------------------
+
+The ports file can be used to specify multiple source and destination pairs that are not sequential in nature.  They can be lids or guids.  If guids, use the -G option to indicate that.
```
Comment (weiny2) [inline doc/rst/common/sec_ports-file.rst]: s/guids/port guids/

File: src/ibtracert.c [hunk 9]
```diff
@@ -759,6 +761,13 @@ static int process_opt(void *context, int ch, char *optarg)
 	switch (ch) {
 	case 1:
 		node_name_map_file = strdup(optarg);
+		if (node_name_map_file == 0)
```
Comment (weiny2) [inline src/ibtracert.c]: This additional check should probably be a separate patch and use NULL for the comparison.  Previous behavior just ignores the error.  Which I agree is probably not what we want.  I wonder what other tools ignore this error...  :-(

File: src/ibtracert.c [hunk 10]
```diff
@@ -759,6 +761,13 @@ static int process_opt(void *context, int ch, char *optarg)
...
+			IBEXIT("out of memory, strdup for node_name_map_file name failed");
+		break;
+	case 2:
+		ports_file = strdup(optarg);
+		if (ports_file == 0)
```
Comment (weiny2) [inline src/ibtracert.c]: NULL

File: src/ibtracert.c [hunk 11]
```diff
@@ -811,9 +826,36 @@ int main(int argc, char **argv)
...
+			IBEXIT("out of memory, strdup for dstid failed");
+     	}
+
+	if (ports_file != NULL) {
+		ports_fd = fopen(ports_file, "r");
+		if (!ports_fd)
+			IBEXIT("cannot open ports-file %s", ports_file);
+        	if (!fgets(ports_buffer, sizeof(ports_buffer), ports_fd)) 
+			IBEXIT("read of ports-file, %s, failed", ports_file);
+		srcid = (char*) malloc(20*sizeof(char));
```
Comment (weiny2) [inline src/ibtracert.c]: Where is this memory freed?  Don't you need to store an array of srcid's and dstid's?  Ok never mind I see it now...  Just have 2 static buffers.

char srcbuf[20];
char dstbuf[20];

And if we are reading the file:

srcid = srcbuf;
dstid = dstbuf;

And why are we only reading the first line of the file?

Can the file contain comments?  ie.

     # this is a ports pair file
     # and some more indented comments

File: src/ibtracert.c (hunk 3 above)
Comment (weiny2) [reply src/ibtracert.c]: Yes need dstid[0] here for correctness.

**Review by weiny2 — COMMENTED** [review]
Over all what you have done here is pretty complicated to review...

I think it would have been better to create a function which resolves src/dest.

foo(src, dst) {
do it...
}

And then if (file) loop through entries in file calling foo.  And if (!file) call foo 1 time.

It would also be easier if it were split into 2 patches.  1 to create foo.  Which is just called 1 time for the same functionality of ibtracert.  And the second one which introduces the file reading code and calls foo 1 time per line in the file.

I would also like to see the file allow for comments because admins like that sort of thing...  ;-)

File: src/ibtracert.c (hunk 4 above)
Comment (skcoulter) [reply src/ibtracert.c]: Are you saying I should add a tab, or you think there is an extra tab ?
I don't see it.

File: src/ibtracert.c (hunk 5 above)
Comment (skcoulter) [reply src/ibtracert.c]: ditto other tab comment

File: src/ibtracert.c (hunk 11 above)
Comment (skcoulter) [reply src/ibtracert.c]: The whole file gets read, see lines 919-935.
Will add the ability to include comments.

File: src/ibtracert.c (hunk 4 above)
Comment (hnrose) [reply src/ibtracert.c]: It looks like there is extra tab from the above as the if is not lined up with the previous line.

File: src/ibtracert.c (hunk 5 above)
Comment (hnrose) [reply src/ibtracert.c]: Again it looks like if is not lined up with line above it.
